core/google/api/core/gapic_v1/method.py

```diff
@@ -0,0 +1,193 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for wrapping low-level gRPC methods with common functionality.
+
+This is used by gapic clients to provide common error mapping, retry, timeout,
+pagination, and long-running operations to gRPC methods.
+"""
+
+import platform
+
+import pkg_resources
+
+from google.api.core import retry
+from google.api.core import timeout
+from google.api.core.helpers import grpc_helpers
+
+_PY_VERSION = platform.python_version()
+_GRPC_VERSION = pkg_resources.get_distribution('grpcio').version
+_API_CORE_VERSION = pkg_resources.get_distribution('google-cloud-core').version
+METRICS_METADATA_KEY = 'x-goog-api-client'
+
+
+def _is_not_none_or_false(value):
+    return value is not None and value is not False
+
+
+def _apply_decorators(func, decorators):
+    """Apply a list of decorators to a given function.
+
+    ``decorators`` may contain items that are ``None`` or ``False`` which will
+    be ignored.
+    """
+    decorators = list(filter(_is_not_none_or_false, decorators))
+
+    for decorator in reversed(decorators):
+        func = decorator(func)
+
+    return func
+
+
+def _prepare_metadata(metadata):
+    """Transforms metadata to gRPC format and adds global metrics.
+
+    Args:
+        metadata (Optional[Mapping[str, str]]): Any current metadata.
+
+    Returns:
+        Sequence[Tuple(str, str)]: The gRPC-friendly metadata keys and values.
+    """
+    if metadata is None:
+        metadata = {}
+
+    client_metadata = 'api-core/{} gl-python/{} grpc/{}'.format(
+        _API_CORE_VERSION, _PY_VERSION, _API_CORE_VERSION)
+
+    # Merge this with any existing metric metadata.
+    client_metadata = ' '.join(filter(None, [
+        client_metadata, metadata.get(METRICS_METADATA_KEY, None)]))
```

Comment:
What does `filter(None, ...)` do?

Reply:
The key might not exist in the metadata, so you could end up with:

```' '.join([client_metadata, None])``` which would raise a TypeError.

So the filter just drops the None from the list so you get:

```
' '.join([client_metadata])
```

Which is just equivalent to `client_metadata`.

If you want, I can re-arrange this as:

```python
if METRICS_METADATA_KEY in metadata:
    client_metadata = '{} {}'.format(client_metadata, metadata[METRICS_METADATA_KEY])
```

Actually - I'm gonna do that anyway.